metaci/release/tasks.py

```diff
@@ -1,16 +1,149 @@
+from collections import defaultdict
 from datetime import datetime, timezone
-from django.dispatch.dispatcher import receiver
+from typing import List
 
-from django_rq import job
 from django.conf import settings
-from django.utils.translation import gettext as _
-from django.urls import reverse
+from django.db.models.query import QuerySet
 from django.db.models.signals import post_delete, post_save
+from django.dispatch.dispatcher import receiver
+from django.urls import reverse
+from django.utils.translation import gettext as _
+from django_rq import job
+from github3.repos.repo import Repository as GitHubRepository
 
+from metaci.build.models import BUILD_STATUSES, Build
+from metaci.plan.models import PlanRepository
 from metaci.release.models import Release, ReleaseCohort
 from metaci.repository.models import Repository
 
-from github3.repos.repo import Repository as GitHubRepository
+
+def _run_planrepo_for_release(release: Release, planrepo: PlanRepository):
+    build = Build(
+        repo=release.repo,
+        plan=planrepo.plan,
+        planrepo=planrepo,
+        commit=release.created_from_commit,
+        build_type="auto",
+        release=release,
+        release_relationship_type="automation",
+    )
+    build.save()
+    if release.status != Release.STATUS.inprogress:
+        release.status = Release.STATUS.inprogress
+        release.save()
+
+
+def find_active_planrepos_by_role(
+    release: Release, role: str
+) -> QuerySet[PlanRepository]:
+    return release.repo.planrepos.filter(plan__role=role, active=True)
+
+
+def fail_release_with_message(release: Release, message: str):
+    release.error_message = _(message)
+    release.status = Release.STATUS.failed
+    release.save()
+
+    release.release_cohort.error_message = _("One or more releases failed.")
+    release.release_cohort.status = Release.STATUS.failed
+    release.release_cohort.save()
+
+
+def _run_release_builds(release: Release):
+    FAILED_STATUSES = [BUILD_STATUSES.error, BUILD_STATUSES.fail]
+    COMPLETED_STATUSES = [BUILD_STATUSES.success, *FAILED_STATUSES]
+
+    # NOTE: bug where MetaCI builds can hang in In Progress forever
+    # will be painful here.
+    def running(builds: List[Build]) -> bool:
+        return bool(builds) and not all(b.status in COMPLETED_STATUSES for b in builds)
```

Comment:
Could be (but no strong opinion from me):
```suggestion
        return any(b.status not in COMPLETED_STATUSES for b in builds)
```